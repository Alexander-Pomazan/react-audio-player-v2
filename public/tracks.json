[
  {
    "id": "269897f9-fb5b-46d0-9933-45104d28123b",
    "name": "Крепость",
    "artist": "Дайте танк",
    "artworkUrl": "/artwork/krepost.jpg",
    "duration": 197,
    "source": "/tracks/dayte-tank-krepost.mp3"
  },
  {
    "id": "9d5196a6-c096-4ca1-a7a7-43192d522245",
    "name": "Люди",
    "artist": "Дайте танк",
    "artworkUrl": "/artwork/ludi.jpeg",
    "duration": "163",
    "source": "/tracks/dayte-tank-ludi.mp3"
  },
  {
    "id": "671363b4-7274-4842-bada-e047a7bf90e3",
    "name": "Шанс",
    "artist": "Дайте танк",
    "artworkUrl": "/artwork/shans.jpeg",
    "duration": 209,
    "source": "/tracks/dayte-tank-shans.mp3"
  },
  {
    "id": "09e49d13-daa1-44a2-8e05-0f4467db9153",
    "name": "Buddy",
    "artist": "Bensound",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/1.jpg?alt=media\u0026token=dd506e0d-2a74-47e9-ad92-2af44c2374c1",
    "duration": 122,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/1.mp3?alt=media\u0026token=485dbea1-03b0-453c-bfa5-2eebba4b7c1f"
  },
  {
    "id": "aaabc5f6-021f-4b7a-abe3-6b17af24165f",
    "name": "A New Beginning",
    "artist": "The Clats",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/2.jpg?alt=media\u0026token=ff778056-8632-45b0-9cd1-3238b2ba0f42",
    "duration": 154,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/2.mp3?alt=media\u0026token=d095ba8c-472f-433e-84d5-4612a5f3abd1"
  },
  {
    "id": "c791a76c-3c04-4385-846e-fd9a36d96010",
    "name": "Happy Rock",
    "artist": "Nora",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/3.jpg?alt=media\u0026token=ab8d335a-c120-41de-80c6-0c00217aef71",
    "duration": 105,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/3.mp3?alt=media\u0026token=6e823a54-2d5e-457d-a386-c5d78e775ec6"
  },
  {
    "id": "5eeb00d7-2133-4591-aa28-20d6f035964f",
    "name": "Ukulele",
    "artist": "Lirae Two",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/4.jpg?alt=media\u0026token=f46a5e83-fa49-44bd-ab71-098890ee3f22",
    "duration": 146,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/4.mp3?alt=media\u0026token=08a390bd-915b-49af-aa1f-ae3099d9ef0e"
  },
  {
    "id": "99c63df1-ca46-4e0b-943b-fc98f885fb5b",
    "name": "Jazzy Frenchy",
    "artist": "Claw",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/5.jpg?alt=media\u0026token=bbd2f5b5-b752-4106-aa6c-824b85e85ec1",
    "duration": 104,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/5.mp3?alt=media\u0026token=c0bd0454-8e99-452f-acb5-ad17456e5f71"
  },
  {
    "id": "cb778e59-1308-47f7-99cb-3ee825ebe3d3",
    "name": "Better Day",
    "artist": "John Doe",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/6.jpg?alt=media\u0026token=91389ee3-8001-4e6c-8688-b63522ff3fb6",
    "duration": 153,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/6.mp3?alt=media\u0026token=447b6068-acb6-424b-a4c8-d2ef3b6ca67e"
  },
  {
    "id": "17d9c0af-8586-4d16-a0d5-fffbc48f0c16",
    "name": "Tomorrow",
    "artist": "Dave Int",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/7.jpg?alt=media\u0026token=d333b9d7-d9aa-46f8-bff0-cd37cf84a412",
    "duration": 294,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/7.mp3?alt=media\u0026token=1626ce91-31c7-40cc-9422-872da0669275"
  },
  {
    "id": "1067ed2b-6c5e-489a-afb5-d83962b27522",
    "name": "Rumble",
    "artist": "Darkgrey",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/8.jpg?alt=media\u0026token=723a91a4-f506-4045-bad5-acf2a1f62e37",
    "duration": 154,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/8.mp3?alt=media\u0026token=0927d467-1dd1-4808-8c97-9badb045c9c9"
  },
  {
    "id": "bc77c7a9-decc-4148-90d4-1a2b677e8898",
    "name": "New Dawn",
    "artist": "Darkgrey",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/9.jpg?alt=media\u0026token=379fdc08-5389-4a13-aa96-b4a733080bb1",
    "duration": 313,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/9.mp3?alt=media\u0026token=ed160936-de07-4012-96dd-e1331732bdc2"
  },
  {
    "id": "3cd2b170-4f3f-4a94-9c3f-585d5f0f1c48",
    "name": "Country Boy",
    "artist": "Dob Bylan",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/10.jpg?alt=media\u0026token=23a2a249-573c-49c4-a249-374b46fee258",
    "duration": 207,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/10.mp3?alt=media\u0026token=bfe7212d-526a-49c0-934f-9aee232bfb39"
  }
]
