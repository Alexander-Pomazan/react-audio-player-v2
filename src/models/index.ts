export * from './track'
export * from './player-status'
export * from './progress'
