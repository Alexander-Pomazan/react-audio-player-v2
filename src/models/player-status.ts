export type PlayerStatus = 'play' | 'pause'
