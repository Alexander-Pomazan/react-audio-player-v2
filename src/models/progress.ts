import { invariant } from 'ts-invariant'

/**
 * From 0 to 1
 */
export type Progress = number

export const progressInvariant = (progress: Progress) => {
  invariant(
    progress >= 0 || progress <= 1,
    'Detected progress value outside of range [0,1]',
  )
}
