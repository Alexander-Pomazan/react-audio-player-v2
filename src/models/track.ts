import { Id } from './id'

export interface Track {
  id: Id
  name: string
  artist: string
  artworkUrl: string
  duration: number
  source: string
}
