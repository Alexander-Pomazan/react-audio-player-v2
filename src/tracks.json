[
  {
    "id": 1,
    "name": "Buddy",
    "artist": "Bensound",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/1.jpg?alt=media\u0026token=dd506e0d-2a74-47e9-ad92-2af44c2374c1",
    "duration": 122,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/1.mp3?alt=media\u0026token=485dbea1-03b0-453c-bfa5-2eebba4b7c1f"
  },
  {
    "id": 2,
    "name": "A New Beginning",
    "artist": "The Clats",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/2.jpg?alt=media\u0026token=ff778056-8632-45b0-9cd1-3238b2ba0f42",
    "duration": 154,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/2.mp3?alt=media\u0026token=d095ba8c-472f-433e-84d5-4612a5f3abd1"
  },
  {
    "id": 3,
    "name": "Happy Rock",
    "artist": "Nora",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/3.jpg?alt=media\u0026token=ab8d335a-c120-41de-80c6-0c00217aef71",
    "duration": 105,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/3.mp3?alt=media\u0026token=6e823a54-2d5e-457d-a386-c5d78e775ec6"
  },
  {
    "id": 4,
    "name": "Ukulele",
    "artist": "Lirae Two",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/4.jpg?alt=media\u0026token=f46a5e83-fa49-44bd-ab71-098890ee3f22",
    "duration": 146,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/4.mp3?alt=media\u0026token=08a390bd-915b-49af-aa1f-ae3099d9ef0e"
  },
  {
    "id": 5,
    "name": "Jazzy Frenchy",
    "artist": "Claw",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/5.jpg?alt=media\u0026token=bbd2f5b5-b752-4106-aa6c-824b85e85ec1",
    "duration": 104,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/5.mp3?alt=media\u0026token=c0bd0454-8e99-452f-acb5-ad17456e5f71"
  },
  {
    "id": 6,
    "name": "Better Day",
    "artist": "John Doe",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/6.jpg?alt=media\u0026token=91389ee3-8001-4e6c-8688-b63522ff3fb6",
    "duration": 153,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/6.mp3?alt=media\u0026token=447b6068-acb6-424b-a4c8-d2ef3b6ca67e"
  },
  {
    "id": 7,
    "name": "Tomorrow",
    "artist": "Dave Int",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/7.jpg?alt=media\u0026token=d333b9d7-d9aa-46f8-bff0-cd37cf84a412",
    "duration": 294,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/7.mp3?alt=media\u0026token=1626ce91-31c7-40cc-9422-872da0669275"
  },
  {
    "id": 8,
    "name": "Rumble",
    "artist": "Darkgrey",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/8.jpg?alt=media\u0026token=723a91a4-f506-4045-bad5-acf2a1f62e37",
    "duration": 154,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/8.mp3?alt=media\u0026token=0927d467-1dd1-4808-8c97-9badb045c9c9"
  },
  {
    "id": 9,
    "name": "New Dawn",
    "artist": "Darkgrey",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/9.jpg?alt=media\u0026token=379fdc08-5389-4a13-aa96-b4a733080bb1",
    "duration": 313,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/9.mp3?alt=media\u0026token=ed160936-de07-4012-96dd-e1331732bdc2"
  },
  {
    "id": 10,
    "name": "Country Boy",
    "artist": "Dob Bylan",
    "artworkUrl": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/10.jpg?alt=media\u0026token=23a2a249-573c-49c4-a249-374b46fee258",
    "duration": 207,
    "source": "https://firebasestorage.googleapis.com/v0/b/react-audio-player-28dec.appspot.com/o/10.mp3?alt=media\u0026token=bfe7212d-526a-49c0-934f-9aee232bfb39"
  }
]
