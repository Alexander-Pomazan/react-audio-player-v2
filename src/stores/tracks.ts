import { atom } from 'jotai'
import { useQuery } from 'react-query'

import { Track } from 'src/models'

const getTracks = async (): Promise<Track[]> => {
  const response = await fetch('/tracks.json')

  return await response.json()
}

export const tracksAtom = atom<Track[]>([])

export const useTracksQuery = () => {
  const { data = [], ...rest } = useQuery('tracks', getTracks, {
    initialData: [],
  })

  return { data, ...rest }
}
