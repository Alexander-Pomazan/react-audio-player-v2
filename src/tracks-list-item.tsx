import React from 'react'
import styled from 'styled-components'

import { Img } from 'src/ui'

// We only want keyboard navigation styles appear when users uses keyboard navigation
// https://stackoverflow.com/questions/31402576/enable-focus-only-on-keyboard-use-or-tab-press
const InnerRoot = styled.div.attrs({ tabIndex: -1 })`
  height: 100%;
  width: 100%;

  display: flex;
  align-items: center;
  border-radius: 2px;

  :focus {
    outline: none;
  }
`

const Root = styled.button<{ isSelected: boolean }>`
  --active-bg-color: #f2f2f2;

  display: block;
  font-size: 1rem;
  height: 100%;
  width: 100%;

  padding: 0;
  background: none;
  border: none;

  cursor: pointer;

  ${(p) => {
    if (p.isSelected)
      return `
      background-color: var(--active-bg-color);
    `
  }}

  :hover {
    background-color: var(--active-bg-color);
  }

  :focus {
    outline: none;
  }

  :focus > ${InnerRoot} {
    position: relative;
    z-index: 10;

    box-shadow: 0 0 0px 1px var(--focus-color);
  }
`

const ArtworkWrapper = styled.div`
  width: 96px;
  height: 80px;

  flex-shrink: 0;
`

const Artwork = styled(Img).attrs({ renderType: 'presentation' })<{
  imageUrl?: string
}>`
  width: 100%;
  height: 100%;

  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
`

const ArtistName = styled.span`
  font-weight: bold;
`

const NamingsWrapper = styled.div`
  margin-left: 2rem;

  font-size: 1.2em;

  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  width: 800px;
  max-width: 100%;
  text-align: left;
`

const Duration = styled.div`
  margin-left: auto;
  margin-right: 1.5em;
`

interface TracksListItemProps {
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void
  isSelected?: boolean
  artist: string
  artworkUrl: string
  trackName: string
  duration: string
}

export const TracksListItem: React.FC<TracksListItemProps> = ({
  artist,
  artworkUrl,
  trackName,
  duration,
  onClick,
  isSelected = false,
}) => {
  return (
    <Root tabIndex={0} onClick={onClick} isSelected={isSelected}>
      <InnerRoot>
        <ArtworkWrapper>
          <Artwork src={artworkUrl} />
        </ArtworkWrapper>

        <NamingsWrapper>
          <ArtistName>{artist}</ArtistName> - {trackName}
        </NamingsWrapper>

        <Duration>{duration}</Duration>
      </InnerRoot>
    </Root>
  )
}
