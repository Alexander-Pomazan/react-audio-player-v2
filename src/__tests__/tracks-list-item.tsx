import 'jest-styled-components'
import styled from 'styled-components'
import React from 'react'
import { render } from '@testing-library/react'
import { TracksListItem } from '../tracks-list-item'
import { Track } from 'src/models/track'
import tracks from 'src/tracks.json'

const [track] = tracks as Track[]

const Button = styled.button`
  color: red;
`

describe('songs-list', () => {
  it('renders', () => {
    const f = render(<Button />)
    f.debug()
    const { debug } = render(
      <TracksListItem {...track} onSelect={console.log} isPlaying={false} />,
    )

    debug()
    expect(1).toBe(1)
  })
})
