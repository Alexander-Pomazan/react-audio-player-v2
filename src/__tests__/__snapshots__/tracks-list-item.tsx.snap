// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`songs-list renders 1`] = `
.c0 {
  color: blue;
}

<div>
  <button
    class="c0"
  />
</div>
`;
